### PR TITLE
[FIX] account: When loading chart template, don't unlink shared records

### PR DESCRIPTION
**Pull request by antoine162**
When loading a chart template on a root company, we check whether any journal items have already been created for the company, and if not, we delete the existing accounting configuration before loading the new CoA.

However, we must take care in the case of records that are shared between several companies, to just unlink them from the active company, rather than delete them (which would affect other companies as well).

This commit fixes that.

task-none

### CODE REVIEW

**Comment by antoine162** [comment]
can be merged already, but to unlink companies from shared accounts without error, we need  https://github.com/odoo/odoo/pull/183695 that redoes the constraint.

File: addons/account/models/chart_template.py [hunk 1]
```diff
@@ -200,11 +200,20 @@ def _load(self, template_code, company, install_demo):
...
                     company_field = 'company_id' if 'company_id' in self.env[model] else 'company_ids'
-                    self.env[model].sudo().with_context(active_test=False).search([(company_field, 'child_of', company.id)]).with_context({MODULE_UNINSTALL_FLAG: True}).unlink()
+                    records = self.env[model].sudo().with_context(active_test=False).search([(company_field, 'child_of', company.id)])
+                    if company_field == 'company_ids':
+                        records_to_keep = records.filtered(lambda r: r.company_ids - child_companies)
+                        records -= records_to_keep
+                        for companies_to_unlink, records_for_companies in records_to_keep.grouped(lambda r: r.company_ids & child_companies).items():
+                            if companies_to_unlink:
+                                # Remove companies from shared records, but don't unlink them
+                                records_for_companies.write({'company_ids': [Command.unlink(c.id) for c in companies_to_unlink]})
```
[the remark this answers is not in the document]
Comment (antoine162) [reply addons/account/models/chart_template.py]: The only issue is that if try to do a `Command.unlink()` for a record that isn't on the `One2many`, you get an error :smile:

[the remark this answers is not in the document]
Comment (william-andre) [reply addons/account/models/chart_template.py]: :hankey:

File: addons/account/models/chart_template.py [hunk 2]
```diff
@@ -201,10 +201,19 @@ def _load(self, template_code, company, install_demo):
...
                     company_field = 'company_id' if 'company_id' in self.env[model] else 'company_ids'
-                    self.env[model].sudo().with_context(active_test=False).search([(company_field, 'child_of', company.id)]).with_context({MODULE_UNINSTALL_FLAG: True}).unlink()
+                    records = self.env[model].sudo().with_context(active_test=False).search([(company_field, 'child_of', company.id)])
+                    if company_field == 'company_ids':
+                        records_to_keep = records.filtered(lambda r: r.company_ids - children_companies)
+                        records -= records_to_keep
+                        for companies_to_unlink, records_for_companies in records_to_keep.grouped(lambda r: r.company_ids & children_companies).items():
+                            if companies_to_unlink:
+                                # Remove companies from shared records, but don't unlink them
+                                records_for_companies.write({'company_ids': [Command.unlink(c.id) for c in companies_to_unlink]})
```
Comment (william-andre) [inline addons/account/models/chart_template.py:215]: Ok if you need the groupby then, but this should work
```suggestion
                        for records_for_companies in records_to_keep.grouped('company_ids').values():
                            records_for_companies.company_ids -= children_companies
```

Comment (antoine162) [reply addons/account/models/chart_template.py]: This does work, thanks so much! :smile:

**Review by william-andre — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**